base/headers.go

```diff
@@ -227,17 +233,115 @@ func (uri WildcardUri) Equals(other Uri) bool {
 }
 
 // Generic list of parameters on a header.
-type Params map[string]MaybeString
+type Params struct {
+	params     map[string]MaybeString
+	paramOrder []string
+}
+
+// Create an empty set of parameters.
+func NewParams() *Params {
+	return &Params{map[string]MaybeString{}, []string{}}
+}
+
+// Returns the entire parameter map.
+func (p *Params) Params() map[string]MaybeString {
+	return p.params
+}
+
+// Returns the requested parameter value.
+func (p *Params) Get(k string) (MaybeString, bool) {
+	v, ok := p.params[k]
+	return v, ok
+}
+
+func (p *Params) Add(k string, v MaybeString) *Params {
+	if p.params == nil {
```

Comment:
If we make Params private, and access it via an interface Params, we can guarantee the caller always uses MakeParams to create it, so we can ensure the members are always set up correctly. So suggest you do that, and then remove this code.